### PR TITLE
Use Subscribed Products instead of Roles for MHR/PPR

### PR DESCRIPTION
**Pull request by dimak1**
*Issue #:* /bcgov/entity#12436

*Description of changes:*
- Refactor to use Products instead of Roles
- Add new and update existing tests
- Cleanup tech debt

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the PPR license (Apache 2.0).

### CODE REVIEW

**Comment by dimak1** [comment]
There could be a few improvements, like adding a state prop for active Subscribed Products codes only (i.e. ['PPR', 'MHR']). But wanted to send it off for initial review. 
@cameron-freshworks @doug-lovett

**Comment by dimak1** [comment]
@cameron-freshworks @doug-lovett marked as ready for review.

File: ppr-ui/src/utils/products-helper.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * @returns String to describe user's roles and products. e.g. STAFF_PPR
+ */
+export function getRoleProductCode (
+  authRoles: Array<string>,
+  subscribedProducts: Array<UserProductSubscriptionIF>): string {
+  let role: string
+
+  if (authRoles.includes('staff') || authRoles.includes('sbc') ||
+    authRoles.includes('helpdesk') || authRoles.includes('ppr_staff') ||
+    authRoles.includes('mhr_staff')) role = 'STAFF'
```
Comment (cameron-eyds) [inline ppr-ui/src/utils/products-helper.ts]: I believe this role is now obsolete (`mhr_staff`)

File: ppr-ui/src/views/Dashboard.vue [hunk 2]
```diff
@@ -431,6 +431,20 @@ export default class Dashboard extends Vue {
...
+      .includes(ProductCode.PPR)
+  }
+
+  private get hasMHRProduct (): boolean {
+    return this.getUserProductSubscriptions
+      .filter(product => product.subscriptionStatus === ProductStatus.ACTIVE)
+      .map(product => product.code)
+      .includes(ProductCode.MHR)
+  }
+
```
Comment (cameron-eyds) [inline ppr-ui/src/views/Dashboard.vue]: Do you think these would be useful globally?

**Review by doug-lovett — APPROVED** [review]
Scanned the change, looks good.

**Review by cameron-eyds — COMMENTED** [review]
Looks good so far! 

Just a few minor questions..
Do we still consider purely the roles for this text templating? As Staff will not have Products.
Also is the intention here to use roles ONLY for the text templating specifically? 

ie we use the roles (+ feature flag) for feature accessibility, the ability to do MHR Searches or start MHR Registrations, for clients we will need to consider products + roles and staff only roles, is that the idea?

Maybe we should have a quick discussion, just hmu whenever 👍

File: ppr-ui/src/utils/products-helper.ts (hunk 1 above)
Comment (dimak1) [reply ppr-ui/src/utils/products-helper.ts]: Removed.

File: ppr-ui/src/views/Dashboard.vue (hunk 2 above)
Comment (dimak1) [reply ppr-ui/src/views/Dashboard.vue]: Added active Product Codes to the store.

File: ppr-ui/src/App.vue [hunk 3]
```diff
@@ -401,6 +405,18 @@ export default class App extends Mixins(AuthMixin) {
...
+      const subscribedProducts = await fetchAccountProducts(this.getAccountId)
+      if (subscribedProducts) {
+        this.setUserProductSubscriptions(subscribedProducts)
+
+        const activeProductCodes = subscribedProducts
+          .filter(product => product.subscriptionStatus === ProductStatus.ACTIVE)
+          .map(product => product.code)
+        this.setUserProductSubscriptionsCodes(activeProductCodes)
+      } else {
+        throw new Error('Unable to get Products for the User')
```
Comment (cameron-eyds) [inline ppr-ui/src/App.vue]: Will this always throw for staff?

File: ppr-ui/src/store/actions/actions-model.ts [hunk 4]
```diff
@@ -44,6 +44,14 @@ export const setAccountProductSubscribtion: ActionIF = (
...
 }
 
+export const setUserProductSubscriptions: ActionIF = ({ commit }, products: Array<UserProductSubscriptionIF>): void => {
+  commit('mutateUserProductSubscriptions', products)
+}
+
+export const setUserProductSubscriptionsCodes: ActionIF = ({ commit }, activeProducts: Array<ProductCode>): void => {
+  commit('mutateUserProductSubscriptionsCodes', activeProducts)
+}
+
```
Comment (cameron-eyds) [inline ppr-ui/src/store/actions/actions-model.ts:54]: Is the latter the abbreviated version of the products? For easier access?

**Review by cameron-eyds — COMMENTED** [review]
Looks good, nice job! 

Just curious, in the event of a staff user (ie no account id or products) that they still have access based on their token roles? Is that still the case?

File: ppr-ui/src/store/actions/actions-model.ts (hunk 4 above)
Comment (dimak1) [reply ppr-ui/src/store/actions/actions-model.ts:54]: Yes, correct. The latter is will hold only the Active Product Codes (for easier access) vs. All Subscribed Products (Active+Inactive).

**Comment by dimak1** [comment]
@cameron-freshworks thanks for the review! I missed that part when Staff will not have an Account Id (it wasn't mentioned). I will update it accordingly and push a new commit.

File: ppr-ui/src/App.vue (hunk 3 above)
Comment (dimak1) [reply ppr-ui/src/App.vue]: Will need to update it to take into consideration that Staff might not have an account id.

Comment (cameron-eyds) [reply ppr-ui/src/App.vue]: Yessir good call, i don't believe they do. They have `BCOL_STAFF_PARTY_CODE` afaik, SBC staff might have an account, not 100% sure. 
But yes, you are correct, we will need to consider staff in these conditions.

File: ppr-ui/src/store/actions/actions-model.ts (hunk 4 above)
Comment (cameron-eyds) [reply ppr-ui/src/store/actions/actions-model.ts:54]: Very cool, thanks!

**Comment by cameron-eyds** [comment]
> @cameron-freshworks thanks for the review! I missed that part when Staff will not have an Account Id (it wasn't mentioned). I will update it accordingly and push a new commit.

No prob, right on!

File: ppr-ui/src/App.vue (hunk 3 above)
Comment (dimak1) [reply ppr-ui/src/App.vue]: Added a check for `accountId`. If a user doesn't have an id (ie staff) then don't fetch the products.

Comment (cameron-eyds) [reply ppr-ui/src/App.vue]: Excellent 👍

**Review by cameron-eyds — APPROVED** [review]
Looks good! 
Due to our pre-release Dev code freeze, hold off on a merge until Divya gives the green light (Thurs i believe)

**Review by doug-lovett — APPROVED** [review]
That's a lot of work, thanks Dima and Cameron for the feedback. I think I followed it.

**Pull request merged**